### ISSUE TITLE
opensc sc_reset leaves unreleased locks with pcsc-lite

### ISSUE THREAD

**Issue by rickyepoderi** [OpenSC/OpenSC#480]
Hi,

This is the same issue explained in #475 but generic. Now we have understood that the problem is general and it does not depend in any card (DNIe in the previous bug).

The problem is that "sc_reset" (when used with pcsc-lite) has a lock leak every time is called. When opensc is configured in shared mode (the default) the issue makes the application to lock the card as it was configured in exclusive mode. If sc_reset is called the card is locked until the process finishes (dosconnects the card and releases the context).

I have prepared a little test program that is in my repo:

https://github.com/rickyepoderi/OpenSC/blob/dnie-lock/src/tools/opensc-error.c

This program performs the following:

1. Create opensc context.
2. Connect to the card and lock it (copied from util.c).
3. Sleep 10s.
4. Request the card atr.
5. Reset the card.
6. Sleep 10s.
7. Unlock the card.
8. Sleep 10s.
9. Disconnect and release.

I am going to present two cases: pcsc-lite with the issue and windows (internal windows pcsc) with another behavior.

a) Linux with pcsc-lite.

Compile the little sample against default debian opensc 0.15.0-1:

    gcc -o opensc-error -I/home/ricky/Desktop/opensc-0.15.0/src/ opensc-error.c /usr/lib/x86_64-linux-gnu/libopensc.so.3

Execute the resulting program two times, the second one is executed after the first one perform the reset (during the second sleep). This is the output of the two programs ordered by time.

    |              EXECUTION 1                            |                EXECUTION 2                          |
    | start: Sat Jun 13 12:43:32 2015                     |                                                     |
    | context created: Sat Jun 13 12:43:34 2015           |                                                     |
    | connect and lock adquired: Sat Jun 13 12:43:34 2015 |                                                     |
    | reset done: Sat Jun 13 12:43:44 2015                |                                                     |
    |                                                     | start: Sat Jun 13 12:43:47 2015                     |
    | lock released: Sat Jun 13 12:43:54 2015             |                                                     |
    | disconnected: Sat Jun 13 12:44:04 2015              |                                                     |
    |                                                     | context created: Sat Jun 13 12:44:04 2015           |
    |                                                     | connect and lock adquired: Sat Jun 13 12:44:04 2015 |
    |                                                     | reset done: Sat Jun 13 12:44:15 2015                |
    |                                                     | lock released: Sat Jun 13 12:44:25 2015             |
    |                                                     | disconnected: Sat Jun 13 12:44:35 2015              |

b) Windows with integrated PCSC.

I compiled OpenSC and then my little program like this:

    cl /Ic:\Users\ricky\Desktop\OpenSC-dnie-lock\src opensc-error.c c:\Users\ricky\Desktop\OpenSC-dnie-lock\src\libopensc\opensc.lib

If I execute the test program two times with the same waiting (the second is started after the reset of the first execution). The output is like this.

    |              EXECUTION 1                            |                EXECUTION 2                          |
    | start: Sat Jun 13 12:22:50 2015                     |                                                     |
    | context created: Sat Jun 13 12:22:50 2015           |                                                     |
    | connect and lock adquired: Sat Jun 13 12:22:50 2015 |                                                     |
    | reset done: Sat Jun 13 12:23:00 2015                |                                                     |
    |                                                     | start: Sat Jun 13 12:23:02 2015                     |
    | lock released: Sat Jun 13 12:23:10 2015             |                                                     |
    |                                                     | context created: Sat Jun 13 12:23:10 2015           |
    |                                                     | connect and lock adquired: Sat Jun 13 12:23:10 2015 |
    | disconnected: Sat Jun 13 12:23:20 2015              |                                                     |
    |                                                     | reset done: Sat Jun 13 12:23:21 2015                |
    |                                                     | lock released: Sat Jun 13 12:23:31 2015             |
    |                                                     | disconnected: Sat Jun 13 12:23:41 2015              |

As you can see in Linux (pcsc-lite) the second execution has to wait until the first execution is completely finished (the disconnect is executed). Nevertheless in windows the second execution starts as soon as the unlock is executed. 

I hope now the problem is much more easy to test and understand. Remember I did a little patch that comments out the extra lock in pcsc-lite (check the previous issue #475).

Thanks in advance!

**Comment by dengert** [comment]
Nice program.

I have some concerns about how OpenSC does locks and reset. Loks look OK, but the reset
appears to be over kill.

Locks appear to be tracked by three different ways:

card->lock_count is incremented by sc_lock, and decremented by sc_unlock

In sc_unlock, if card->lock_count == 0, pcsc_lock will be called,
in sc_unlock if card->lock_count == 1, pcsc_unlock will be called.

in reader-pcsc.c
pcsc_lock will call SCardBeginTransaction and set reader->priv->locked = 1
pcsc_unlock will call SCardEndTransaction and set reader->priv->locked = 0

pcscd has two types of locks, the ones defined by SCardBeginTransaction/SCardEndTransaction operatons
and the shared/exclusive set by the SCardConnect/SCardDisconnect

Both SCardEndTransaction and SCardDisconnect offer SCARD_RESET_CARD

The problem I am having, is why pcsc_reset uses the pcsc_reconnect (that tries to use the SCardDisconnect, then SCardConnect)
rather then use then SCardEndTransaction? This looks like overkill, and there is a lot for extra
code that may be holding on to the lock when it should not.

IMHO The pcsc_reset use of old_locked is questionable.

Can you modify your program to write to the OpenSC debug log, so we can see the debug statements for all the locking?
(Have the opensc lof write to stdout, or have two different opensc.conf files one for each process, so they use different
debug files.)
This might require printing the card->lock_count and reader->priv->locked,
as well as add some more PCSC_TRACE to show SCardBeginTransaction/SCardEndTransaction, and SCardConnect/SCardDisconnect
so these would show on Windows as well an linux.


On 6/13/2015 6:28 AM, rickyepoderi wrote:
> Hi,
>
> This is the same issue explained in #475 <https://github.com/OpenSC/OpenSC/issues/475> but generic. Now we have understood that the problem is general and it does not depend in any card (DNIe in the
> previous bug).
>
> The problem is that "sc_reset" (when used with pcsc-lite) has a lock leak every time is called. When opensc is configured in shared mode (the default) the issue makes the application to lock the card
> as it was configured in exclusive mode. If sc_reset is called the card is locked until the process finishes (dosconnects the card and releases the context).
>
> I have prepared a little test program that is in my repo:
>
> https://github.com/rickyepoderi/OpenSC/blob/dnie-lock/src/tools/opensc-error.c
>
> This program performs the following:
>
>  1. Create opensc context.
>  2. Connect to the card and lock it (copied from util.c).
>  3. Sleep 10s.
>  4. Request the card atr.
>  5. Reset the card.
>  6. Sleep 10s.
>  7. Unlock the card.
>  8. Sleep 10s.
>  9. Disconnect and release.
>
> I am going to present two cases: pcsc-lite with the issue and windows (internal windows pcsc) with another behavior.
>
> a) Linux with pcsc-lite.
>
> Compile the little sample against default debian opensc 0.15.0-1:
>
> |gcc -o opensc-error -I/home/ricky/Desktop/opensc-0.15.0/src/ opensc-error.c /usr/lib/x86_64-linux-gnu/libopensc.so.3
> |
>
> Execute the resulting program two times, the second one is executed after the first one perform the reset (during the second sleep). This is the output of the two programs ordered by time.
>
> ||              EXECUTION 1                            |                EXECUTION 2                          |
> | start: Sat Jun 13 12:43:32 2015                     |                                                     |
> | context created: Sat Jun 13 12:43:34 2015           |                                                     |
> | connect and lock adquired: Sat Jun 13 12:43:34 2015 |                                                     |
> | reset done: Sat Jun 13 12:43:44 2015                |                                                     |
> |                                                     | start: Sat Jun 13 12:43:47 2015                     |
> | lock released: Sat Jun 13 12:43:54 2015             |                                                     |
> | disconnected: Sat Jun 13 12:44:04 2015              |                                                     |
> |                                                     | context created: Sat Jun 13 12:44:04 2015           |
> |                                                     | connect and lock adquired: Sat Jun 13 12:44:04 2015 |
> |                                                     | reset done: Sat Jun 13 12:44:15 2015                |
> |                                                     | lock released: Sat Jun 13 12:44:25 2015             |
> |                                                     | disconnected: Sat Jun 13 12:44:35 2015              |
> |
>
> b) Windows with integrated PCSC.
>
> I compiled OpenSC and then my little program like this:
>
> |cl /Ic:\Users\ricky\Desktop\OpenSC-dnie-lock\src opensc-error.c c:\Users\ricky\Desktop\OpenSC-dnie-lock\src\libopensc\opensc.lib
> |
>
> If I execute the test program two times with the same waiting (the second is started after the reset of the first execution). The output is like this.
>
> ||              EXECUTION 1                            |                EXECUTION 2                          |
> | start: Sat Jun 13 12:22:50 2015                     |                                                     |
> | context created: Sat Jun 13 12:22:50 2015           |                                                     |
> | connect and lock adquired: Sat Jun 13 12:22:50 2015 |                                                     |
> | reset done: Sat Jun 13 12:23:00 2015                |                                                     |
> |                                                     | start: Sat Jun 13 12:23:02 2015                     |
> | lock released: Sat Jun 13 12:23:10 2015             |                                                     |
> |                                                     | context created: Sat Jun 13 12:23:10 2015           |
> |                                                     | connect and lock adquired: Sat Jun 13 12:23:10 2015 |
> | disconnected: Sat Jun 13 12:23:20 2015              |                                                     |
> |                                                     | reset done: Sat Jun 13 12:23:21 2015                |
> |                                                     | lock released: Sat Jun 13 12:23:31 2015             |
> |                                                     | disconnected: Sat Jun 13 12:23:41 2015              |
> |
>
> As you can see in Linux (pcsc-lite) the second execution has to wait until the first execution is completely finished (the disconnect is executed). Nevertheless in windows the second execution starts
> as soon as the unlock is executed.
>
> I hope now the problem is much more easy to test and understand. Remember I did a little patch that comments out the extra lock in pcsc-lite (check the previous issue #465
> <https://github.com/OpenSC/OpenSC/pull/465>).
>
> Thanks in advance!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/480>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by rickyepoderi** [comment]
Hi Doug,

I'm going to try to summarize again how I see this problem:

* The opensc manages shared locking at several levels:
  * General level at card.c
  * Particular level in reader-pcsc.c for pcsc implementation.
* The problem here is in reader-pcsc.c (particular level with pcsc interaction). So we can forget about general level managed in card.c.
* You can also use the card in exclusive mode but this is another topic, we are using it in shared mode. So please forget that part too.
* In reader-pcsc.c the locking is managed using the SCardBeginTransaction and SCardEndTransaction functions that PCSC defines in the API. Besides there is a internal var that the opensc manages to know if it is locked or not. This var is just a boolean (0 not locked, 1 locked) which is stored in the private data of the reader (priv->locked).
* This mechanism works more or less ok (in both opensc and pcsc-lite) but there is a special case which is the PCSC function SCardReconnect. This function in Windows PCSC implementation releases the locks at PCSC level. For that reason opensc acts accordingly. In line 429 of pcsc-reader (function pcsc_reconnect) the locked mark is set to 0 (priv->locked = 0) and in line 583 of the pcsc_reset the lock is assigned again because of that reason.
* The real problem is that pcsc-lite (Linux implementation of PCSC) does not release any lock in SCardReconnect. So in Linux the statement that SCardReconnect unlocks the card is false and opensc is assigning an extra lock that blocks the card forever.

The problem is indeed in pcsc-lite (this Linux implementation does not work as the windows PCSC in SCardReconnect), but Ludovic thought is easier to change opensc for the moment. 

What do you want me to do with the debug part? Do you want me to put the opensc.conf debug var to 9 and execute the program again? Should I copy the output for both executions here?

Remember that my little program works for any card (not DNIe exclusively) so you can test with any card you have. If you use opensc with pcsc-lite the problem will arise.

Thanks!

**Comment by dengert** [comment]
Other OpenSC developers need to get involved with this ticket.

OK, you have a solution, but say below "Ludovic thought is easier to change opensc for the moment."
If OpenSC adds your patch but the problem is really in pcsc-lite, how do we remove the patch when
pcsc-lite is fixed?
How do we keep them in sync?

One way is to use a flag in the opensc.conf "reader_driver pcsc" section to say use your circumvention for now.
and only users who have the problem would set the new flag.

For you situation, have you tried setting: disconnect_action = unpower;

Does the DNIe really need a reset?
As I said in earlier e-mail only two cards call the sc_reset(card,0);
why do they need this and other cards don't?


On 6/14/2015 8:24 AM, rickyepoderi wrote:
> Hi Doug,
>
> I'm going to try to summarize again how I see this problem:
>
>   * The opensc manages shared locking at several levels:
>       o General level at card.c
>       o Particular level in reader-pcsc.c for pcsc implementation.
>   * The problem here is in reader-pcsc.c (particular level with pcsc interaction). So we can forget about general level managed in card.c.
>   * You can also use the card in exclusive mode but this is another topic, we are using it in shared mode. So please forget that part too.
>   * In reader-pcsc.c the locking is managed using the SCardBeginTransaction and SCardEndTransaction functions that PCSC defines in the API. Besides there is a internal var that the opensc manages to
>     know if it is locked or not. This var is just a boolean (0 not locked, 1 locked) which is stored in the private data of the reader (priv->locked).
>   * This mechanism works more or less ok (in both opensc and pcsc-lite) but there is a special case which is the PCSC function SCardReconnect. This function in Windows PCSC implementation releases the
>     locks at PCSC level. For that reason opensc acts accordingly. In line 429 of pcsc-reader (function pcsc_reconnect) the locked mark is set to 0 (priv->locked = 0) and in line 583 of the pcsc_reset
>     the lock is assigned again because of that reason.
>   * The real problem is that pcsc-lite (Linux implementation of PCSC) does not release any lock in SCardReconnect. So in Linux the statement that SCardReconnect unlocks the card is false and opensc is
>     assigning an extra lock that blocks the card forever.
>
> The problem is indeed in pcsc-lite (this Linux implementation does not work as the windows PCSC in SCardReconnect), but Ludovic thought is easier to change opensc for the moment.
>
> What do you want me to do with the debug part? Do you want me to put the opensc.conf debug var to 9 and execute the program again? Should I copy the output for both executions here?
>
> Remember that my little program works for any card (not DNIe exclusively) so you can test with any card you have. If you use opensc with pcsc-lite the problem will arise.
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/480#issuecomment-111827005>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by LudovicRousseau** [comment]
You have no way from the PC/SC API to know the behavior of SCardDisconnect(). Will it remove any lock or not? you cant tell.

Microsoft WinSCard and pcsc-lite differ on this point. I have not checked what Apple has to say about this.

I documented the difference at http://pcsclite.alioth.debian.org/api/group__API.html#details point 6. I do not plan to change pcsc-lite. See also the bug I opened at pcsc-lite https://alioth.debian.org/tracker/index.php?func=detail&aid=315083&group_id=30105&atid=410088

**Comment by rickyepoderi** [comment]
If you want I can test DNIe without the reset, I can comment the line in the cwa14890.c and try it out. I will do it when I have some time. I will come back here with the results.

But Doug, you have to understand that sc_reset is there, defined in opensc.h, and anyone can use it in any card. Right now it is buggy in combination with pcsc-lite and the patch we are proposing is absolutely minimal (do the same in windows and do not change locks in other circumstances). The fact that only two cards are using sc_reset is the reason of being unaware of the problem until now.

**Comment by dengert** [comment]
I understand sc_reset is buggy and only two cards appear to use it.
No one has spoken up as to why sc_reset is needed in these two cards.


Since Ludovic says he does not plan to change pcsc-lite, then your change in OpenSC maybe the best way for now.


On 6/15/2015 7:57 AM, rickyepoderi wrote:
> If you want I can test DNIe without the reset, I can comment the line in the cwa14890.c and try it out. I will do it when I have some time. I will come back here with the results.
>
> But Doug, you have to understand that sc_reset is there, defined in opensc.h, and anyone can use it in any card. Right now it is buggy in combination with pcsc-lite and the patch we are proposing is
> absolutely minimal (do the same in windows and do not change locks in other circumstances). The fact that only two cards are using sc_reset is the reason of being unaware of the problem until now.
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/480#issuecomment-112059093>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by rickyepoderi** [comment]
Yep! I don't know if the sc_reset is really needed or why it is there in DNIe. Moreover yesterday night I did some tests and the card works properly without the reset (I connected to the demo page and then I signed with the applet). Everything went ok with the DNIe with no reset in cwa14890.c.

I have checked and the sc_reset is used in all the DNIe implementations (the first one done directly done by Spanish government and the current implementation in opensc). Previously I thought that the sc_reset was there to clean the card state before starting another trusted channel (a kind of reset any previous channel or card disposition cos a new trusted channel is going to be established). Maybe it is just an extra security step in order to start the channel afresh.

In the other card (which I think is card-mcrd.c) the sc_reset is used to move the card to a known state, at least it is commented like this. I suppose the card is reset when some strange RC is returned just to start it properly.

I don't know, this issue is very weird. Doug, how do we continue? If you want more tests with the DNIe just tell me.

Thanks!

**Comment by dengert** [comment]
@martinpaljak can you give some background on these changes?

On 2010-11-30 commit 7ab591a6840f7017d673300a5262305e0ee2d1e0 shows the do_cold_reset parameter was added sc_reset and pcsc_reset was modified
to reconnect with either SCARD_UNPOWER_CARD or SCARD_RESET_CARD This was only used by the card-mcrd.c

What problem did it solve for card-mcrd.c?


On 2010-11-29 commit c1c3aa1d55c176a0c37d71263ad568b0fd6486f4

PC/SC: make (dis)connect actions configurable, SCardDisconnect, SCardEndTransaction and SCardReconnect
     actions can now be configured via opensc.conf in better detail.


On 6/16/2015 2:34 AM, rickyepoderi wrote:
> Yep! I don't know if the sc_reset is really needed or why it is there in DNIe. Moreover yesterday night I did some tests and the card works properly without the reset (I connected to the demo page and
> then I signed with the applet). Everything went ok with the DNIe with no reset in cwa14890.c.
>
> I have checked and the sc_reset is used in all the DNIe implementations (the first one done directly done by Spanish government and the current implementation in opensc). Previously I thought that the
> sc_reset was there to clean the card state before starting another trusted channel (a kind of reset any previous channel or card disposition cos a new trusted channel is going to be established).
> Maybe it is just an extra security step in order to start the channel afresh.
>
> In the other card (which I think is card-mcrd.c) the sc_reset is used to move the card to a known state, at least it is commented like this. I suppose the card is reset when some strange RC is
> returned just to start it properly.
>
> I don't know, this issue is very weird. Doug, how do we continue? If you want more tests with the DNIe just tell me.
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/480#issuecomment-112321595>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by martinpaljak** [comment]
- The purpose of sc_reset is to reset the card (power-cycle, cold or warm, meaning a direct connection to reader-pcsc.c no matter what card.c and OpenSC locking thinks, in theory it might try to re-establish the transaction-backd lock ) without depending on the presence of transactions (thus the reconnect instead of transactions)
- Correct, the use in card-mcrd.c is to work around a seriously buggy card.
- Maybe there are implementation issues in sc_reset behaviour that need to be fixed.
- My guess is that the use in DNIe driver is bogus with its use of sc_reset.

**Comment by dengert** [comment]
@germanblanco  and Juan,

can you comment on the sc_reset in the original cwa14890.c code
and why it says:
  /* reset card (warm reset, do not unpower card) */

Is the need for a warm reset documented in any DNIe or CWA documents?

What happens if the card is unpowered at this point?

On 6/17/2015 10:08 AM, Martin Paljak wrote:
>   * The purpose of sc_reset is to reset the card (power-cycle, cold or warm, meaning a direct connection to reader-pcsc.c no matter what card.c and OpenSC locking thinks, in theory it might try to
>     re-establish the transaction-backd lock ) without depending on the presence of transactions (thus the reconnect instead of transactions)
>   * Correct, the use in card-mcrd.c is to work around a seriously buggy card.

OK, good to know this was done for a specific card problem.
Are these cards still in use?

>   * Maybe there are implementation issues in sc_reset behaviour that need to be fixed.

Yes, and this sounds like @rickyepoderi has a patch for this, but we should understand
why these cards need a warm reset.

This bug report is about different behavior of sc_reset on linux and windows.
Does the mcrd card have this same problem?

>   * My guess is that the use in DNIe driver is bogus with its use of sc_reset.

@rickyepoderi depending on the answers from Jaun and @germanblanco can you try and and uses sc_reset(card, 1) in the DNIe code, and see if that fixes the problem?
Does it work without the sc_reset?

>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/480#issuecomment-112837080>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by rickyepoderi** [comment]
Hi Martin,

The problem with pcsc_reset / pcsc_reconnect is that with pcsc-lite the transactions are not released. For that reason any call to pcsc_reset is adding an extra lock which (in the case of DNIe) locks the card until that process closes the connection.

So, we have two problems here:

* pcsc_reset / pcsc_reconnect are not handling locks correctly with pcsc-lite.
* The call to sc_reset in DNIe maybe is not needed.

I think we have to face the two problems separately. I prefer that this bug will continue with the first problem. If you want we can open another bug with the DNIe and the unneeded call to sc_reset.

Thanks!

**Comment by miguel-cv** [comment]
The FNMT (spanish dni manufacturer) has released the sources of the "official" driver at https://www.sede.fnmt.gob.es/documents/11614/4531260/MultiPKCS11_v1_3_0.rar . I don't know if it's good to put it here as it has no licence file, but i only want to say that there is no reset in the secure channel operation (or i don't see it).

**Comment by rickyepoderi** [comment]
Hi again,

@dengert "sc_reset(card, 1)" has the same problem (tested right now) and it is quite logical. The problem is with lock management and not with the type of reset. The lock is not released in the same way and the card remains locked until the process finishes and diconnects.

I have checked this new implementation (it is nice to have another one). There are two interesting things:

* The file "Sources/Comm_DNIe.cpp" has the same logic to initialize the trusted channel (CComm_DNIe::setChannel) and you are right, it seems that no reset is done there. So maybe the reset is definitely not needed.

* The second thing is how they reconnect (in file "Sources/Slot.cpp" the method "CSlot::Connect"), the are doing this:

        #if defined _YOSEMITE_

                SCardDisconnect( m_hCard, SCARD_RESET_CARD );

                lReturn = SCardConnect
                                        (
                                                g_hContext,
                                                m_sName.c_str(),
                                                dwShareMode,
                                                SCARD_PROTOCOL_T0,
                                                &m_hCard,
                                                &protocol
                                        );
                #else
                lReturn = SCardReconnect
                (
                    m_hCard,
                    dwShareMode,
                    SCARD_PROTOCOL_T0,
                    SCARD_RESET_CARD,
                    &protocol
                );
        #endif

So if it is macos  SCardDisconnect + SCardConnect is done but if it is windows just SCardReconnect is used (I supposse Linux is not even tested). So those guys found exactly the same problem we are having in Linux because I think that macos winscard implementation is based on pcsc-lite. If you check the code of the same file ("Sources/Slot.cpp") the lock is added after the connect just as opensc is doing now, for example line 690:

                                         Connect();
 
                                         if(transactionBegun)
                                                 BeginTransaction();

But remember that they are doing a SCardDisconnect + SCardConnect instead of SCardReconnect because of the issue we are commenting here.

As I said before, I prefer to deal with the two issues separately. 

@dengert Why don't we open another bug with DNIe sc_reset? This way this issue is just for fixing the pcsc_reset problem.

@LudovicRousseau What do you think about the solution of the FNMT people? I still think that it is better to use reconnect but not re-adding the locks, but obviously doing a SCardDisconnect + SCardConnect is also a solution.

**Comment by germanblanco** [comment]
Hello,
I don't know the reasons for that reset, but by the way it looks, I would clearly vote for removing it from DNIe.
This leaves an apparently buggy sc_reset function that is used only in one card, so I would also vote for checking why and how that card uses it before working on fixing anything else.

**Comment by LudovicRousseau** [comment]
Mac OS X Yosemite has a bug with SCardReconnect(). See http://ludovicrousseau.blogspot.fr/2014/12/os-x-yosemite-bug-scardreconnect_16.html

I think that explains the special case in the FNMT code.

**Comment by rickyepoderi** [comment]
This is just my idea. We have two separate issues: locking problem in SCardReconnect / SCardReset; a possible un-needed call to sc_reset in DNIe.

This issue (please check again the title) is about the first problem. I think we have to open another issue for the second problem.

My opinion is that fixing the issue commented in this thread is important. Any card that is using or is going to use sc_reset or sc_reconnect is affected by this problem (right now DNIe and mcrd). At the moment we have two possible fixes:

* Call to SCardReconnect but then you have to re-add the lock or not depending you are using windows or pcsc-lite.
* Change pcsc_reconnect calling SCardReconnect in windows and SCardDisconnect + SCardConnect in pcsc-lite (as FNMT did).

The first solution makes the connection work without loosing the lock (if pcsc-lite) but it change the final result of pcsc_reconnect depending your OS (in windows after a pcsc_reconnect the card is not locked and in pcsc-lite it is still locked). The second solution makes pcsc_reconnect to act exactly the same way in windows and linux.

Regards!

**Comment by rickyepoderi** [comment]
I have implemented a little example of the second option (SCardDisconnect + SCardConnect). It works with the little test program executed for the outputs presented in the first comment (the output in Linux now is  the same than in Windows) and with the DNIe sample test page.

Remember that I implemented a tentative patch for the first option in the bug #475.

I have realized that sc_reconnect does not exists in opensc.h (only sc_reset is there) so, if sc_reconnect is not going to be added to the .h file, I prefer the first option.

Thanks!

**Comment by dengert** [comment]
You have heard from the 3 developers that know the code: @LudovicRousseau, @martinpaljak and @germanblanco, follow their advise.
If you and they think that sc_reconnect should be added to OpenSC, then add it.
Originally it was not clear how your change would affect other cards and that was my main concern. It sounds like it will not, so I don't have a problem.

The only question I have is with the OPenSC DNIe, and why cwa calls sc_reset. Someone added that call to OpenSC, but the FNMT code does not use it.
The only thing I imagine for the use of a warm reset, is to have the card partially reset something, for example
a previous verify of the pin, stays in effect, but the SM SSC is reset or the need to use SM is reset. (or maybe its the other way around.)
But the FNMT does not do a reset, so I would agree with @germanblanco and vote to remove it.

I do not have a DNIe or msrd card or any card that needs sc_reset. So its up to you and other developers to make the decisions.


On 6/19/2015 5:38 AM, rickyepoderi wrote:
> I have implemented a little example of the second option (SCardDisconnect + SCardConnect). It works with the little test program executed for the outputs presented in the first comment (the output in
> Linux now is the same than in Windows) and with the DNIe sample test page.
>
> Remember that I implemented a tentative patch for the first option in the bug #475 <https://github.com/OpenSC/OpenSC/issues/475>.
>
> I have realized that sc_reconnect does not exists in opensc.h (only sc_reset is there) so, if sc_reconnect is not going to be added to the .h file, I prefer the first option.
>
> Thanks!
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/480#issuecomment-113467167>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by rickyepoderi** [comment]
Hi Doug,

I have submitted the pull request. Finally I decided to go with the first option, if sc_reconnect is not exported in opensc.h I think this solution is better (although the behavior of pcsc_reconnect is different in windows than in linux it does not matter cos this method cannot be used, the sc_reset behavior is exactly the same in both environments).

I have tested the solution in Linux and in windows with DNIe card. If you have any doubt please just comment me.

Thanks!

**Comment by gargsurbhi18** [comment]
In my application the scenario is 
1. i connect the reader and card using pcscd
2.   after sometime reader has been unplugged. 
3. next i againg plugged in the reader.
4. i disconnect and resetcontext but it is giving error
5. then i am not able to connect the card again.
please help me in resolving the issue

**Comment by frankmorgner** [comment]
We recently fixed some race condition https://github.com/OpenSC/OpenSC/commit/d449feeda6b3d8ec27a184417d9b5d1ba73df6d7
@gargsurbhi18 , If you have trouble even with that patch applied, the please open a **new issue**.
